src/persist-client/src/write.rs

```diff
@@ -101,40 +115,221 @@ where
+            "WriteHandle::append timeout={:?} new_upper={:?}",
+            timeout,
+            new_upper
+        );
+        let deadline = Instant::now() + timeout;
+
+        let lower = self.upper.clone();
+        let upper = new_upper;
+        let since = Antichain::from_elem(T::minimum());
+        let desc = Description::new(lower, upper, since);
+
+        // TODO: Instead construct a Vec of blob keys here so it can be empty
+        // (if there are no updates) and bounded memory usage (if updates is
+        // large).
+        let key = Uuid::new_v4().to_string();
+        let mut value = Vec::new();
+        if let Err(err) = Self::encode_batch(&mut value, &desc, updates) {
+            return Ok(Err(err));
+        }
+        self.blob
+            .set(deadline, &key, value, Atomicity::RequireAtomic)
+            .await?;
+
+        let res = self
+            .machine
+            .append(deadline, &self.writer_id, &[key], &desc)
+            .await?;
+        match res {
+            Ok(_) => self.upper = desc.upper().clone(),
+            Err(err) => return Ok(Err(err)),
+        };
+        Ok(Ok(()))
+    }
+
+    /// Applies `updates` to this shard and downgrades this handle's upper to
+    /// `new_upper` iff the current shard upper is equal to `expected_upper`.
+    ///
+    /// In contrast to [Self::append], this linearizes mutations from all
+    /// writers. It's intended for use as an atomic primitive for timestamp
+    /// bindings, SQL tables, etc.
+    ///
+    /// All times in `updates` must be greater or equal to `expected_upper` and
+    /// not greater or equal to `new_upper`. A `new_upper` of the empty
+    /// antichain "finishes" this shard, promising that no more data is ever
+    /// incoming.
+    ///
+    /// `updates` may be empty, which allows for downgrading `upper` to
+    /// communicate progress. It is unexpected to call this with `new_upper`
+    /// equal to `self.upper()`, as it would mean `updates` must be empty
+    /// (making the entire call a no-op).
+    ///
+    /// This uses a bounded amount of memory, even when `updates` is very large.
+    /// Individual records, however, should be small enough that we can
+    /// reasonably chunk them up: O(KB) is definitely fine, O(MB) come talk to
+    /// us.
+    ///
+    /// The clunky two-level Result is to enable more obvious error handling in
+    /// the caller. See <http://sled.rs/errors.html> for details.
+    pub async fn compare_and_append<'a, I: IntoIterator<Item = ((&'a K, &'a V), &'a T, &'a D)>>(
+        &mut self,
+        timeout: Duration,
+        updates: I,
+        expected_upper: Antichain<T>,
+        new_upper: Antichain<T>,
     ) -> Result<Result<(), InvalidUsage>, LocationError> {
         trace!(
             "WriteHandle::write_batch timeout={:?} new_upper={:?}",
             timeout,
             new_upper
         );
+        let deadline = Instant::now() + timeout;
+
+        let lower = expected_upper;
```

Comment:
The only diff to `append()` is how we determine `lower` here and that we call `compare_and_append()` below. Maybe we can factor out some common code. Fine if you already thought about it and found it too much of a hassle.

Reply:
I think you were right but not anymore XD! I think with the 2 vs 3 level result, the common code extracted wouldn't really get us much